### PR TITLE
Fix LayoutLMv2 init issue and doctest

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

(I think I have done all I can do for doctest - there are 2 other failures where I ping other people)

`LayoutLMv2Model.visual_segment_embedding`'s initialization (when it is not in a checkpoint) won't be reproducible even if seed is set. This is because it is defined

> self.visual_segment_embedding = nn.Parameter(nn.Embedding(1, config.hidden_size).weight[0])

and we have `_fast_init` used in loading +`_init_weights` is not designed to deal with `nn.Parameter`.

This PR fixes the reproducibility issue by initializing `visual_segment_embedding` explicitly.

Remark: such places requiring this fix is very few - so we don't destroy the purpose of having `_fast_init`.

### CODE REVIEW

File: src/transformers/models/layoutlmv2/modeling_layoutlmv2.py [hunk 1]
```diff
@@ -697,6 +697,10 @@ def __init__(self, config):
         self.visual_proj = nn.Linear(config.image_feature_pool_shape[-1], config.hidden_size)
         if self.has_visual_segment_embedding:
             self.visual_segment_embedding = nn.Parameter(nn.Embedding(1, config.hidden_size).weight[0])
+            # For `nn.Parameter`, we need to give it an initialized weight in order to keep reproducibility.
+            # Otherwise, since `_fast_init` is used in `modeling_utils.py`, the layer `nn.Embedding` is not initialized,
+            # and `_init_weights` is not designed to deal with `nn.Parameter`.
+            self.visual_segment_embedding.data.normal_(mean=0.0, std=self.config.initializer_range)
```
Comment (ydshieh) [inline src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:703]: this is the main fix

**Comment by ydshieh** [comment]
The failure in ` tests_pr_documentation_tests` is just missing `detectron2`. I can install it for that job.

File: src/transformers/models/layoutlmv2/modeling_layoutlmv2.py [hunk 2]
```diff
@@ -697,6 +697,10 @@ def __init__(self, config):
         self.visual_proj = nn.Linear(config.image_feature_pool_shape[-1], config.hidden_size)
         if self.has_visual_segment_embedding:
             self.visual_segment_embedding = nn.Parameter(nn.Embedding(1, config.hidden_size).weight[0])
+            # For `nn.Parameter`, we need to give it an initialized weight in order to keep reproducibility.
+            # Otherwise, since `_fast_init` is used in `modeling_utils.py`, the layer `nn.Embedding` is not initialized,
+            # and `_init_weights` is not designed to deal with `nn.Parameter`.
```
Comment (amyeroberts) [inline src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:702]: I'm not sure why we can't have this logic in `_init_weights` ?

Comment (ydshieh) [reply src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:702]: The definition of `_init_weights` looks like `def _init_weights(self, module):`, where we check things like

> if isinstance(module, nn.Embedding)
> elif isinstance(module, (nn.Linear, nn.Conv1d)
etc.  (using layer type to determine the init method)

We don't check if it is `nn.Parameter` as it is too generic. For example, we also have

```python
self.q_bias = nn.Parameter(torch.zeros(1, 1, self.all_head_size))
self.v_bias = nn.Parameter(torch.zeros(1, 1, self.all_head_size))
```
(where they are initialized with `zeros` , and not treated in `_init_weights`.

However, here it kind special:

> self.visual_segment_embedding = nn.Parameter(nn.Embedding(1, config.hidden_size).weight[0])

it was `nn.Embedding` but wrappered into `nn.Parameter`. I guess it is because we don't want to have bias (?)
But since it is `nn.Parameter`, it is not treated in `nn.Parameter`.

Comment (ydshieh) [reply src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:702]: For some models (say, `VivitPreTrainedModel`), we have ` isinstance(module, nn.Parameter)` in `_init_weights`

```
         elif isinstance(module, nn.Parameter):
            module.data.normal_(mean=0.0, std=self.config.initializer_range)
```
But for `LayoutLM`, it has multiple `nn.Parameter` and they don't mean to be initialized in the same way.

Comment (ydshieh) [reply src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:702]: Gentilly ping @amyeroberts 😃

Comment (ydshieh) [reply src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:702]: @amyeroberts Let me know if you have further question.

Comment (amyeroberts) [reply src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:702]: Yes, sorry for being late on this. I see. My understanding of the init weights behaviour is that if a weight has been initialized then it's marked as such and skipped it it matches other patterns later on (correct me if I'm wrong here). Would it be possible then to instead something similar to what we see in e.g. [wav2vec2's _init_weights](https://github.com/huggingface/transformers/blob/8b02bb6e748842cfb74ec9cba808201b893fa0af/src/transformers/models/wav2vec2/modeling_wav2vec2.py#L1075) where we check if the module is `LayoutLMv2Model` and then init this param? 

I realise it's not a huge difference having this in this _init_ than _init_weights, but there's an increasing amount of logic e.g. with quantization which assumes this behaviour is controlled within the `_init_weights` module

File: src/transformers/models/layoutlmv2/modeling_layoutlmv2.py [hunk 3]
```diff
@@ -822,7 +829,7 @@ def forward(
         >>> import torch
         >>> from datasets import load_dataset
 
-        >>> set_seed(88)
+        >>> set_seed(0)
```
Comment (ydshieh) [inline src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:832]: all these and below have to be updated

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing and iterating!

File: src/transformers/models/layoutlmv2/modeling_layoutlmv2.py (hunk 3 above)
Comment (ydshieh) [reply src/transformers/models/layoutlmv2/modeling_layoutlmv2.py:828]: turns out they work on the CI runners. No need to update
